### DISCUSSION TITLE
More and more Lockin

### DISCUSSION THREAD

**Discussion by AdsonCicilioti** [vercel/next.js#69601] [Help] [answered]
### Summary

Next.js is amazing but it increasingly makes me look at other fremaworks.

After hours of development, I reached the authentication part where when I needed middleware, I discovered that it only works with Edge RUntime (from Vercel). In other words, requiring our applications to host their Apps on Vercel. This led me to try alternatives but that leads me to a very unnecessary level of increased abstraction and maintenance.

If anyone has a solution on how to get working middleware outside of Vercel's Edge Runtime, please let me know.

**Comment by icyJoseph** [comment]
> After hours of development, I reached the authentication part where when I needed middleware, I discovered that it only works with Edge RUntime (from Vercel). In other words, requiring our applications to host their Apps on Vercel. This led me to try alternatives but that leads me to a very unnecessary level of increased abstraction and maintenance.

That's incorrect, but with a catch, that makes it look correct.

I've hosted Next.js apps in various service providers, with middleware, and in those cases, it just runs, side by side, with your app. The catch, is that, even though you run it in your own, while developing, and building, you still have to play by the edge rules, or else...

Like, you can run `next build` and `next start`, on your machine, and things just work right? You don't suddenly find your self not being able to run you app, on your machine.

**Reply by AdsonCicilioti** [reply]
Hi @icyJoseph ! othanks for your answer.

My analysis may be incorrect, but I still haven't figured out how to make `middleware.ts` work outside of the Edge Runtime.
Solutions and libraries exist but I believe if I have to expand the base just to get around this it becomes unfeasible, since there are countless other meta frameworks that basically do the same things. But Next.js with these approaches ends up generating breaking points in adherence to it.

**Reply by icyJoseph** [reply, marked as answer]
> My analysis may be incorrect, but I still haven't figured out how to make `middleware.ts` work outside of the Edge Runtime.

This is what I mean. When developing, building and running, your code has to be compatible with Web APIs, which are typically what the wave of JavaScript runtimes support, rather than Node.js - it has to be less runtime specific if you will.

However, you are not locked in, in terms of deployment. You can deploy anywhere, it'll work. Though, achieving the same feature set as in Vercel can be tricky, but not impossible.

> Solutions and libraries exist but I believe if I have to expand the base just to get around this it becomes unfeasible, since there are countless other meta frameworks that basically do the same things. But Next.js with these approaches ends up generating breaking points in adherence to it.

Here's the issue. Node.js is itself a JavaScript runtime, which in turn made its own set of APIs. Nowadays, we have several runtimes, Vercels Edge, being one of them, being built on top of Cloudflare workers:

> Vercel's Edge Functions are built on top of Cloudflare Workers. The runtime is designed for any similar provider, but we've chosen Cloudflare for now because they have an amazing product. - [source](https://news.ycombinator.com/item?id=29003514)

Take for example, Jose, https://www.npmjs.com/package/jose?activeTab=readme, what it does is possible in Node.js, with the Node.js APIs but those are not portable. Jose in turn, is built on Web APIs, so it works across runtimes.

As usual, there's a lot more context and nuance to these things and opinion leaders are unfortunately, quick to emit judgment and developers are left hanging. Even I'm leaving a lot out of this post right here.

**Reply by AdsonCicilioti** [reply]
@icyJoseph Thank you very much for your explanations. I really needed this clarification. I completely agree with the decoupling of Apis from bad teams like Node/Deno/etc. 

I just hope Vercel doesn't leave Next.js to their RUntime for some "lockin" approach.
